Jam jar playback

ethan-riback opened the pull request:
JamJars are now functional in this branch.

Several video page functionality was also made available, such as video likes, video dislikes, and video flagging.

ethan-riback committed:
Initial research and development for switching between videos.

ethan-riback committed:
Created a StitchedJamJarAVPlayerViewController that derives from the video player. It will be used to create new functionality exclusive to JamJars.
There is a bare JamJarPage. It has the JamJar player working.
Got slider and observers working with JamJars.

ethan-riback committed:
Created AVPlayer subclass that keeps track of the videoId for the video.
JamJarPlayer now will start with the correct video.
Storage for videos and AVPlayers now exists for JamJarPlayer.

ethan-riback committed:
JamJars keep track of when a new video begins to overlap.

ethan-riback committed:
JamJar functionality works. No collection view is present to show possible videos to switch to. Swiping is the only way to switch videos for now.

ethan-riback committed:
Updated the VideoPageView so with like, dislike, and flag buttons.
Created a working view for JamJarPageView. This view does not have a working suggested contents table.
The like, dislike, and flag buttons are also not functional.

ethan-riback committed:
Rewind and Fast forward buttons now have jammer functionality. Suggested content table was added to JamJar page and was fixed for the video page.

ethan-riback committed:
Video votes are now tracked in the video model and work has begun on vote interaction with videos. No server calls to update vote status is functional yet.

ethan-riback committed:
Videos new record when they are viewed.

ethan-riback committed:
Merge branch 'master' into JamJarPlayback

ethan-riback committed:
Small changes to get branch working with the new pushes from Master

ethan-riback committed:
Video Votes work

ethan-riback committed:
Videos can now be flagged!!!! Removed a lot of print statements.

ethan-riback committed:
VideoPlayer on UploadViewController has no full screen option.

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+    
+
+    //Variable to store video frame
+    var jamJarContainerFrameInPortrait: CGRect!
```
Out of curiousity, why do we do this?

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+        
+        // Register the reusable video cell
+        self.suggestedTableView.registerNib(UINib(nibName: "JamJarCell", bundle: nil), forCellReuseIdentifier: "JamJarCell")
+        self.suggestedTableView.registerNib(UINib(nibName: "JamJarHeader", bundle: nil), forCellReuseIdentifier: "JamJarHeader")
```
This needs to be changed to match this https://github.com/projectjamjar/iOSJamJar/blob/master/JamJar/ProfileViewController.swift#L57

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+        self.suggestedTableView.registerNib(UINib(nibName: "JamJarHeader", bundle: nil), forCellReuseIdentifier: "JamJarHeader")
+        
+        // Set Report Reasons Data
+        self.reasonsToFlag = ["Accuracy", "Inappropriate", "Quality","Report User"];
```
Why don't we just declare these above?  They're not changing, so it doesn't really make sense to do it here...

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+    
+    deinit {
+        //When the video page is dismissed, remove observers from the AVPlayerController
+        removeOberservsInPlayer()
```
Spelling - `removeObserversInPlayer`

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+    }
+    
+    // Record a view
+    func recordView(videoId: Int!) {
```
This is totally nitpicky but can we call this "updateViewCount" or something.  The name of this func makes it sound like we're recording a video or something.

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+        self.view.layoutIfNeeded()
+    }
+    
+    func removeOberservsInPlayer() {
```
As noted above, fix this spelling

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+        }
+    }
+    
+    func changeJamJarInPlayer() {
```
What does this function do?  Does it hot-swap a jamjar?  Does it change the video?  I'm confused :/

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+        self.likeButton.selected = (video.userVote == true)
+        self.dislikeButton.selected = (video.userVote == false)
+        self.likeButton.highlightBackgroundLike()
+        self.dislikeButton.highlightBackgroundDislike()
```
Don't forget to trigger a change to the vote count here

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+                    self.updateLikeDislikeButtons()
+                }
+            }
+        } else {
```
This is super code-cloned and can definitely be chopped down

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+                    self.updateLikeDislikeButtons()
+                }
+            }
+        } else {
```
Same with dis

markkohdev commented on JamJar/JamJarPageViewController.swift:
```diff
@@ -0,0 +1,371 @@
+    
+    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
+        if(segue.identifier == "playJamJar") {
+            unowned let embeddedVideoViewController = segue.destinationViewController as! StitchedJamJarAVPlayerViewController
```
We should add some protection here to check if the hls_src exists here so the app doesn't crash if the video is still processing

markkohdev commented on JamJar/StitchedJamJarAVPlayerViewController.swift:
```diff
@@ -0,0 +1,253 @@
+        super.updateTimeLabel(elapsedTime, duration: duration)
+        
+        //TODO: find overlaps
+        let edges = jamjar.nodes![String(self.currentVideo.id!)]
```
Uhhh, this seems like a **lot** of computation to do every time this callback is called.  I feel like there's gotta be some preprocessing that we can do on load here so that we can do a constant-time lookup for any edges.

markkohdev commented on JamJar/StitchedJamJarAVPlayerViewController.swift:
```diff
@@ -0,0 +1,253 @@
+     * The Methods below are used to pull information from arrays that store
+     * information about videos and AVPlayers. I tried getting this to work
+     * for the Array extension but could not and I wanted to focus on getting
+     * JamJars to work. @mark: let's make sure we revisit this.
```
Sounds good, let's chat about it more in-depth tonight.

markkohdev commented on JamJar/VideoPageViewController.swift:
```diff
@@ -48,12 +54,21 @@ class VideoPageViewController: BaseViewController, UITableViewDelegate, UITableV
         artistsLabel.text = video.getArtistsString()
         venueLabel.text = concert.venue.name
         dateLabel.text = concert.date.string("MM-d-YYYY")
+        likesCountLabel.text = String((video.videoVotes.filter{$0.vote == 1}.first?.total)!)
```
This is code-cloned from the jamjar VC I think.  Can you extract this "likesCount" out to the Video class? (and same with the dislikes)

markkohdev commented on JamJar/VideoPageViewController.swift:
```diff
@@ -94,10 +110,142 @@ class VideoPageViewController: BaseViewController, UITableViewDelegate, UITableV
         self.view.layoutIfNeeded()
     }
     
+    func removeOberservsInPlayer() {
```
Fix spelling

markkohdev commented on JamJar/VideoPageViewController.swift:
```diff
@@ -94,10 +110,142 @@ class VideoPageViewController: BaseViewController, UITableViewDelegate, UITableV
+                }
+            }
+        } else {
+            VideoService.vote(video.id, vote: true) { success, error in
```
Alright soooo, we have a whole bunch of code cloned here as well.  Maybe make a base player page class or something?

markkohdev commented on JamJar/VideoPageViewController.swift:
```diff
@@ -94,10 +110,142 @@ class VideoPageViewController: BaseViewController, UITableViewDelegate, UITableV
+        // Add Notes Text Field
+        let notes = flagView.addTextField("Notes")
+        
+        flagView.addButton("Submit") {
```
More cloning than Dolly :sheep: :sheep: :sheep: :sheep: (http://www.animalresearch.info/en/medical-advances/timeline/cloning-dolly-the-sheep/)

markkohdev commented:
@ethan-riback Left a whole buncha comments.  Please fix!

markkohdev committed:
Merge branch 'master' of github.com:projectjamjar/iOSJamJar into JamJarPlayback

markkohdev committed:
Added ChooseVideosController, CustomVideoPicker, ground work for in-app-recording.

markkohdev committed:
Moved video selector logic to choose videos view.  Added thumbnail retriever to video items.

markkohdev committed:
Improved collection view styling and messed around with flow layouts.

ethan-riback committed:
Made the sign up page a scroll view

ethan-riback committed:
Merge remote-tracking branch 'origin/JamJarPlayback' into JamJarPlayback

ethan-riback committed:
Fixed video UI layout issues. Resizing looks good on all iOS devices and does not become weird if the video switches in landscape.

markkohdev committed:
Added ToUploadVideos segue, delete from staging, and some other stuff.

ethan-riback committed:
Worked on some code cleaning and removing some code cloning.

markkohdev committed:
Fixed buttons and stuff.  Made a bunch more stuff work for in-video capture.

markkohdev committed:
Don't add already added videos.

ethan-riback committed:
Removed a lot of code cloning.

markkohdev committed:
Did more customization on jamjar camera.  Fixed a buncha issues.  Autofilled video names with filenames so uploads don't fail.  Other bananas.

markkohdev committed:
Merge branch 'JamJarPlayback' of github.com:projectjamjar/iOSJamJar into feature/in-video-capture

ethan-riback committed:
Set up AVPlayer limit for JamJars.
Fixed bug on Sign Up page.

ethan-riback committed:
Resized VideoPlayer UI.
Made the VideoPlayer UI fade in and out.

ethan-riback committed:
Updated JamJarCell to just display the number of videos on the JamJar.
App no longer has an automatic swipe gesture for going back in a navigation controller.

ethan-riback committed:
Merge pull request #13 from projectjamjar/feature/in-video-capture

Feature/in video capture
